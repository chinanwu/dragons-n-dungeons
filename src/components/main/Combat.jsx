import PropTypes from 'prop-types';
import React, { useCallback, useState } from 'react';
import { createPortal } from 'react-dom';
import { connect } from 'react-redux';

import { enterBtn, spaceBtn } from '../../constants/Keycodes';
import isNumber from '../../functions/isNumber';
import validateEvent from '../../functions/validateEvent';
import { defaultState } from '../../reducers/CombatReducer';
import { applyCombat } from '../../thunk/CombatThunk.jsx';
import Checkbox from '../general/Checkbox.jsx';
import Input from '../general/Input.jsx';
import './Combat.less';

import InputBox from './InputBox.jsx';

export const Combat = ({
  inspiration,
  armourClass,
  passivePerception,
  speed,
  hitPoints,
  initiativeBonus,
  proficiencyBonus,
  hitDice,
  tempHitPoints,
  onChange,
}) => {
  const [dynamicHitPoints, setDynamicHitPoints] = useState(hitPoints);
  const [showHitPointsTotalModal, setShowHitPointsTotalModal] = useState(false);

  const handleChange = type =>
    useCallback(
      event => {
        if (event && event.target) {
          if (event.target.value) {
            onChange(type, event.target.value.toString());
          } else {
            onChange(type, defaultState[type]);
          }
        }
      },
      [onChange]
    );

  const handleKeyDown = type =>
    useCallback(num => onChange(type, num), [onChange]);

  const handleDynamicHitPointsChange = useCallback(event => {
    if (event && event.target) {
      if (event.target.value) {
        const value = isNumber(event.target.value);
        value ? setDynamicHitPoints(value) : null;
      } else {
        setDynamicHitPoints(defaultState['hitPoints']);
      }
    }
  }, []);

  const handleDynamicHitPointsKeyDown = useCallback(
    num => {
      setDynamicHitPoints(num);
    },
    [setDynamicHitPoints]
  );

  const handleHitPointsClick = callback =>
    useCallback(() => {
      setDynamicHitPoints(dynamicHitPoints => callback(dynamicHitPoints));
    }, [setDynamicHitPoints, dynamicHitPoints]);

  const handleHitPointsKeyDown = callback =>
    useCallback(
      event => {
        if (validateEvent(event)) {
          if (
            !event.shiftKey &&
            !event.ctrlKey &&
            !event.altKey &&
            !event.metaKey
          ) {
            if (event.keyCode) {
              if (event.keyCode === enterBtn || event.keyCode === spaceBtn) {
                event.preventDefault();
                setDynamicHitPoints(dynamicHitPoints =>
                  callback(dynamicHitPoints)
                );
              }
            }
          }
        }
      },
      [setDynamicHitPoints, dynamicHitPoints]
    );

  const handleHitPointsTotalClick = useCallback(() => {
    setShowHitPointsTotalModal(!showHitPointsTotalModal);
  }, [showHitPointsTotalModal, setShowHitPointsTotalModal]);

  // https://fontawesome.com/license
  // SVG below is from font awesome

  return (
    <div className="Combat">
      <div className="Combat__top">
        <div className="Combat__deathSaves">
          <div className="Combat__label">Death Saves</div>
          <div className="Combat__deathSavesContent">
            <div id="deathSaveFailLabel" className="Combat__deathSaveFailLabel">
              Fails
            </div>
            <div
              id="deathSaveSuccessLabel"
              className="Combat__deathSaveSuccessLabel"
            >
              Successes
            </div>
            <div className="Combat__deathSavesFailCheckboxes">
              <Checkbox
                id="deathSaveFail-1"
                ariaLabelledBy="deathSaveFailLabel"
              />
              <Checkbox
                id="deathSaveFail-2"
                ariaLabelledBy="deathSaveFailLabel"
              />
              <Checkbox
                id="deathSaveFail-3"
                ariaLabelledBy="deathSaveFailLabel"
              />
            </div>
            <div className="Combat__deathSavesSuccessCheckboxes">
              <Checkbox
                id="deathSaveSuccess-1"
                ariaLabelledBy="deathSaveSuccessLabel"
              />
              <Checkbox
                id="deathSaveSuccess-2"
                ariaLabelledBy="deathSaveSuccessLabel"
              />
              <Checkbox
                id="deathSaveSuccess-3"
                ariaLabelledBy="deathSaveSuccessLabel"
              />
            </div>
          </div>
        </div>
        <div className="Combat__inspiration">
          <div id="inspirationLabel" className="Combat__label">
            Inspiration
          </div>
          <div className="Combat__inspirationCheckbox">
            <Checkbox
              id="inspirationCheckbox"
              ariaLabelledBy="inspirationLabel"
              checked={inspiration}
            />
          </div>
        </div>
        <div className="Combat__armourClass">
          <InputBox
            id="armourClass"
            className="Combat__input--larger"
            label="Armour Class"
            value={armourClass}
            type="text"
            onChange={handleChange('armourClass')}
          />
        </div>
      </div>
      <div className="Combat__middle">
        <div className="Combat__middleSide">
          <div className="Combat__passivePerception">
            <InputBox
              id="passivePerception"
              className="Combat__input--larger"
              label="Passive Perception"
              value={passivePerception}
              type="number"
              onKeyDown={handleKeyDown('passivePerception')}
              onChange={handleChange('passivePerception')}
            />
          </div>
          <div className="Combat__speed">
            <InputBox
              id="speed"
              className="Combat__input--larger"
              label="Speed"
              value={speed}
              type="number"
              onKeyDown={handleKeyDown('speed')}
              onChange={handleChange('speed')}
            />
          </div>
        </div>
        <div className="Combat__hitPoints">
          <div className="Combat__hitPointsHeader">
            <div
              className="Combat__hitPointsBtn"
              role="button"
              tabIndex="0"
              unselectable="on" //IE9 downwards and Opera
              onKeyDown={handleHitPointsKeyDown(num => num + 1)}
              onClick={handleHitPointsClick(num => num + 1)}
            >
              +
            </div>
            <div id="hitPointsLabel" className="Combat__label">
              Hit Points
            </div>
            <div
              className="Combat__hitPointsBtn"
              role="button"
              tabIndex="0"
              unselectable="on" //IE9 downwards and Opera
              onKeyDown={handleHitPointsKeyDown(num =>
                num - 1 < 0 ? 0 : num - 1
              )}
              onClick={handleHitPointsClick(num => (num - 1 < 0 ? 0 : num - 1))}
            >
              -
            </div>
          </div>
          <div className="Combat__hitPointsInput">
            <Input
              id="hitPoints"
              className="Combat__input--largest"
              type="number"
              value={dynamicHitPoints}
              ariaLabelledBy="hitPointsLabel"
              onKeyDown={handleDynamicHitPointsKeyDown}
              onChange={handleDynamicHitPointsChange}
            />
          </div>
          <div className="Combat__hitPointsTotal">
            <div id="hitsPointsTotalLabel">Total:</div>
            <Input
              id="hitPoints"
              className="Combat__hitPointsTotalInput"
              type="number"
              value={hitPoints}
              ariaLabelledBy="hitsPointsTotalLabel"
              onKeyDown={handleKeyDown('hitPoints')}
              onChange={handleChange('hitPoints')}
            />
            <button
              className="Combat__hitPointsTotalEditBtn"
              onClick={handleHitPointsTotalClick}
            >
              <svg
                aria-hidden="true"
                focusable="true"
                data-prefix="fas"
                data-icon="pencil-alt"
                className="svg-inline--fa fa-pencil-alt fa-w-16 Combat__hitPointsTotalEditBtnSvg"
                role="img"
                xmlns="http://www.w3.org/2000/svg"
                viewBox="0 0 512 512"
                height={12}
              >
                <path
                  fill="currentColor"
                  d="M497.9 142.1l-46.1 46.1c-4.7 4.7-12.3 4.7-17 0l-111-111c-4.7-4.7-4.7-12.3 0-17l46.1-46.1c18.7-18.7 49.1-18.7 67.9 0l60.1 60.1c18.8 18.7 18.8 49.1 0 67.9zM284.2 99.8L21.6 362.4.4 483.9c-2.9 16.4 11.4 30.6 27.8 27.8l121.5-21.3 262.6-262.6c4.7-4.7 4.7-12.3 0-17l-111-111c-4.8-4.7-12.4-4.7-17.1 0zM124.1 339.9c-5.5-5.5-5.5-14.3 0-19.8l154-154c5.5-5.5 14.3-5.5 19.8 0s5.5 14.3 0 19.8l-154 154c-5.5 5.5-14.3 5.5-19.8 0zM88 424h48v36.3l-64.5 11.3-31.1-31.1L51.7 376H88v48z"
                />
              </svg>
            </button>
            {showHitPointsTotalModal &&
              createPortal(<div>test</div>, document.body)}
          </div>
        </div>
        <div className="Combat__middleSide">
          <div className="Combat__initiativeBonus">
            <InputBox
              id="initiativeBonus"
              className="Combat__input--larger"
              label="Initiative Bonus"
              value={initiativeBonus}
              type="number"
              onKeyDown={handleKeyDown('initiativeBonus')}
              onChange={handleChange('initiativeBonus')}
            />
          </div>
          <div className="Combat__proficiencyBonus">
            <InputBox
              id="proficiencyBonus"
              className="Combat__input--larger"
              label="Proficiency Bonus"
              value={proficiencyBonus}
              type="number"
              onKeyDown={handleKeyDown('proficiencyBonus')}
              onChange={handleChange('proficiencyBonus')}
            />
          </div>
        </div>
      </div>
      <div className="Combat__bottom">
        <div className="Combat__hitDice">
          <InputBox
            id="hitDice"
            className="Combat__input--larger"
            label="Hit Dice"
            value={hitDice}
            type="text"
            onChange={handleChange('hitDice')}
          />
        </div>
        <div className="Combat__tempHitPoints">
          <InputBox
            id="tempHitPoints"
            className="Combat__input--larger"
            label="Temporary Hit Points"
            value={tempHitPoints}
            type="number"
            onKeyDown={handleKeyDown('tempHitPoints')}
            onChange={handleChange('tempHitPoints')}
          />
        </div>
      </div>
    </div>
  );
};

Combat.propTypes = {
  inspiration: PropTypes.bool,
  armourClass: PropTypes.number,
  passivePerception: PropTypes.number,
  speed: PropTypes.number,
  hitPoints: PropTypes.number,
  initiativeBonus: PropTypes.number,
  proficiencyBonus: PropTypes.number,
  hitDice: PropTypes.string,
  tempHitPoints: PropTypes.number,
  onChange: PropTypes.func,
};

export const mapStateToProps = ({
  combat: {
    inspiration,
    armourClass,
    passivePerception,
    speed,
    hitPoints,
    initiativeBonus,
    proficiencyBonus,
    hitDice,
    tempHitPoints,
  },
}) => ({
  inspiration: inspiration,
  armourClass: armourClass,
  passivePerception: passivePerception,
  speed: speed,
  hitPoints: hitPoints,
  initiativeBonus: initiativeBonus,
  proficiencyBonus: proficiencyBonus,
  hitDice: hitDice,
  tempHitPoints: tempHitPoints,
});

export const mapDispatchToProps = {
  onChange: applyCombat,
};

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(Combat);
