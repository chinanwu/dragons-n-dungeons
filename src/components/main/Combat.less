@import '../styling/colours.less';
@import '../styling/fonts.less';

.Combat {
  display: flex;
  flex-direction: column;
  margin-top: 12px;
  border: 1px solid black;
}

.Combat__top {
  display: grid;
  grid-template-columns: 2fr 1fr 2fr;
  margin-top: 20px;
  margin-left: 40px;
  margin-right: 40px;
}

.Combat__middle {
  display: grid;
  grid-template-columns: 1fr 2fr 1fr;
  margin-top: 12px;
  margin-left: 40px;
  margin-right: 40px;
}

.Combat__middleSide {
  display: grid;
  grid-template-rows: 1fr 1fr;
}

.Combat__bottom {
  display: grid;
  grid-template-columns: 1fr 2fr;
  margin: 12px 40px 20px;
}

.Combat__label {
  text-align: center;
  margin-top: 6px;
}

.Combat__input--larger {
  text-align: center;
  font-size: @font-36;
}

.Combat__input--largest {
  text-align: center;
  font-size: @font-72;
}

.Combat__deathSaves {
  border: 1px solid black;
}

.Combat__deathSavesContent {
  display: grid;
  grid-template-columns: 1fr 1fr;
  grid-template-areas:
    'fail  fCheckbox'
    'success sCheckbox';
  margin-left: 8px;
  margin-right: 8px;
}

.Combat__deathSaveFailLabel {
  grid-area: fail;
}

.Combat__deathSaveSuccessLabel {
  grid-area: success;
}

.Combat__deathSavesFailCheckboxes {
  display: flex;
  grid-area: fCheckbox;
}

.Combat__deathSavesSuccessCheckboxes {
  display: flex;
  grid-area: sCheckbox;
}

.Combat__inspiration {
  display: grid;
  grid-template-rows: 1fr 2fr;
  border: 1px solid black;
  margin-left: 12px;
  margin-right: 12px;
  padding-left: 3px;
  padding-right: 3px;
}

.Combat__inspirationCheckbox {
  display: flex;
  justify-content: center;
}

.Combat__armourClass {
  border: 1px solid black;
}

.Combat__passivePerception {
  border: 1px solid black;
  margin-bottom: 12px;
}

.Combat__speed {
  border: 1px solid black;
}

.Combat__hitPoints {
  border: 1px solid black;
  margin-left: 12px;
  margin-right: 12px;
  display: grid;
  grid-template-rows: 1fr 5fr 1fr;
}

.Combat__hitPointsHeader {
  display: flex;
  justify-content: space-between;
  align-items: center;
  margin-left: 12px;
  margin-right: 12px;
}

.Combat__hitPointsBtn {
  cursor: pointer;

  -webkit-user-select: none; /* Chrome/Safari */
  -moz-user-select: none; /* Firefox */
  -ms-user-select: none; /* IE10+ */
}

.Combat__hitPointsInput {
  display: flex;
  justify-content: center;
  align-items: center;
}

.Combat__hitPointsTotal {
  display: grid;
  grid-template-columns: 9fr 90fr 1fr;
  align-items: center;
  font-size: @font-18;
  margin-bottom: 6px;
  margin-left: 12px;
  margin-right: 20px;
}

.Combat__hitPointsTotalInput {
  margin-left: 3px;
  font-size: @font-18;
}

.Combat__hitPointsTotalEditBtn {
  border: none;
  cursor: pointer;
  padding: 0;
  background-color: inherit;
}

.Combat__hitPointsTotalEditBtnSvg {
  color: @main-5;
}

.Combat__initiativeBonus {
  border: 1px solid black;
  margin-bottom: 12px;
}

.Combat__proficiencyBonus {
  border: 1px solid black;
}

.Combat__hitDice {
  border: 1px solid black;
  margin-right: 12px;
}

.Combat__tempHitPoints {
  border: 1px solid black;
}

@media all and (max-width: 1175px) {
  .Combat {
    margin-top: 0;
    margin-left: 20px;
    margin-right: 10px;
  }

  .Combat__top {
    margin-left: 20px;
    margin-right: 20px;
  }

  .Combat__middle {
    margin-left: 20px;
    margin-right: 20px;
  }

  .Combat__bottom {
    margin-left: 20px;
    margin-right: 20px;
  }
}

@media all and (max-width: 655px) {
  .Combat__deathSavesContent {
    display: flex;
    flex-direction: column;
  }

  .Combat__deathSaveFailLabel {
    order: -1;
    text-align: center;
    margin-top: 3px;
  }

  .Combat__deathSavesFailCheckboxes {
    order: 0;
  }

  .Combat__deathSaveSuccessLabel {
    text-align: center;
    order: 1;
  }

  .Combat__deathSavesSuccessCheckboxes {
    order: 2;
    margin-top: 3px;
    margin-bottom: 6px;
  }
}
