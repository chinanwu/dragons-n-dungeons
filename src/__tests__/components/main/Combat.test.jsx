import { shallow } from 'enzyme';
import React from 'react';

import { Combat } from '../../../components/main/Combat.jsx';

jest.unmock('../../../components/main/Combat.jsx');

describe('Combat component', () => {
  describe('rendering', () => {
    it('renders Combat', () => {
      const wrapper = shallow(<Combat />);
      expect(wrapper.getElement()).toMatchSnapshot();
    });
  });
});
