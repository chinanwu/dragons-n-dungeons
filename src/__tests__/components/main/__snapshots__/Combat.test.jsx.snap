// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Combat component rendering renders Combat 1`] = `
<div
  className="Combat"
>
  <div
    className="Combat__top"
  >
    <div
      className="Combat__deathSaves"
    >
      <div
        className="Combat__label"
      >
        Death Saves
      </div>
      <div
        className="Combat__deathSavesContent"
      >
        <div
          className="Combat__deathSaveFailLabel"
          id="deathSaveFailLabel"
        >
          Fails
        </div>
        <div
          className="Combat__deathSaveSuccessLabel"
          id="deathSaveSuccessLabel"
        >
          Successes
        </div>
        <div
          className="Combat__deathSavesFailCheckboxes"
        >
          <Checkbox
            ariaLabelledBy="deathSaveFailLabel"
            id="deathSaveFail-1"
          />
          <Checkbox
            ariaLabelledBy="deathSaveFailLabel"
            id="deathSaveFail-2"
          />
          <Checkbox
            ariaLabelledBy="deathSaveFailLabel"
            id="deathSaveFail-3"
          />
        </div>
        <div
          className="Combat__deathSavesSuccessCheckboxes"
        >
          <Checkbox
            ariaLabelledBy="deathSaveSuccessLabel"
            id="deathSaveSuccess-1"
          />
          <Checkbox
            ariaLabelledBy="deathSaveSuccessLabel"
            id="deathSaveSuccess-2"
          />
          <Checkbox
            ariaLabelledBy="deathSaveSuccessLabel"
            id="deathSaveSuccess-3"
          />
        </div>
      </div>
    </div>
    <div
      className="Combat__inspiration"
    >
      <div
        className="Combat__label"
        id="inspirationLabel"
      >
        Inspiration
      </div>
      <div
        className="Combat__inspirationCheckbox"
      >
        <Checkbox
          ariaLabelledBy="inspirationLabel"
          id="inspirationCheckbox"
        />
      </div>
    </div>
    <div
      className="Combat__armourClass"
    >
      <InputBox
        className="Combat__input--larger"
        id="armourClass"
        label="Armour Class"
        onChange={[Function]}
        type="text"
      />
    </div>
  </div>
  <div
    className="Combat__middle"
  >
    <div
      className="Combat__middleSide"
    >
      <div
        className="Combat__passivePerception"
      >
        <InputBox
          className="Combat__input--larger"
          id="passivePerception"
          label="Passive Perception"
          onChange={[Function]}
          onKeyDown={[Function]}
          type="number"
        />
      </div>
      <div
        className="Combat__speed"
      >
        <InputBox
          className="Combat__input--larger"
          id="speed"
          label="Speed"
          onChange={[Function]}
          onKeyDown={[Function]}
          type="number"
        />
      </div>
    </div>
    <div
      className="Combat__hitPoints"
    >
      <div
        className="Combat__hitPointsHeader"
      >
        <div
          className="Combat__hitPointsBtn"
          onClick={[Function]}
          onKeyDown={[Function]}
          role="button"
          tabIndex="0"
          unselectable="on"
        >
          +
        </div>
        <div
          className="Combat__label"
          id="hitPointsLabel"
        >
          Hit Points
        </div>
        <div
          className="Combat__hitPointsBtn"
          onClick={[Function]}
          onKeyDown={[Function]}
          role="button"
          tabIndex="0"
          unselectable="on"
        >
          -
        </div>
      </div>
      <div
        className="Combat__hitPointsInput"
      >
        <Input
          ariaLabelledBy="hitPointsLabel"
          className="Combat__input--largest"
          id="hitPoints"
          onChange={[Function]}
          onKeyDown={[Function]}
          type="number"
        />
      </div>
      <div
        className="Combat__hitPointsTotal"
      >
        <div
          id="hitsPointsTotalLabel"
        >
          Total:
        </div>
        <Input
          ariaLabelledBy="hitsPointsTotalLabel"
          className="Combat__hitPointsTotalInput"
          id="hitPoints"
          onChange={[Function]}
          onKeyDown={[Function]}
          type="number"
        />
        <button
          className="Combat__hitPointsTotalEditBtn"
          onClick={[Function]}
        >
          <svg
            aria-hidden="true"
            className="svg-inline--fa fa-pencil-alt fa-w-16 Combat__hitPointsTotalEditBtnSvg"
            data-icon="pencil-alt"
            data-prefix="fas"
            focusable="true"
            height={12}
            role="img"
            viewBox="0 0 512 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M497.9 142.1l-46.1 46.1c-4.7 4.7-12.3 4.7-17 0l-111-111c-4.7-4.7-4.7-12.3 0-17l46.1-46.1c18.7-18.7 49.1-18.7 67.9 0l60.1 60.1c18.8 18.7 18.8 49.1 0 67.9zM284.2 99.8L21.6 362.4.4 483.9c-2.9 16.4 11.4 30.6 27.8 27.8l121.5-21.3 262.6-262.6c4.7-4.7 4.7-12.3 0-17l-111-111c-4.8-4.7-12.4-4.7-17.1 0zM124.1 339.9c-5.5-5.5-5.5-14.3 0-19.8l154-154c5.5-5.5 14.3-5.5 19.8 0s5.5 14.3 0 19.8l-154 154c-5.5 5.5-14.3 5.5-19.8 0zM88 424h48v36.3l-64.5 11.3-31.1-31.1L51.7 376H88v48z"
              fill="currentColor"
            />
          </svg>
        </button>
      </div>
    </div>
    <div
      className="Combat__middleSide"
    >
      <div
        className="Combat__initiativeBonus"
      >
        <InputBox
          className="Combat__input--larger"
          id="initiativeBonus"
          label="Initiative Bonus"
          onChange={[Function]}
          onKeyDown={[Function]}
          type="number"
        />
      </div>
      <div
        className="Combat__proficiencyBonus"
      >
        <InputBox
          className="Combat__input--larger"
          id="proficiencyBonus"
          label="Proficiency Bonus"
          onChange={[Function]}
          onKeyDown={[Function]}
          type="number"
        />
      </div>
    </div>
  </div>
  <div
    className="Combat__bottom"
  >
    <div
      className="Combat__hitDice"
    >
      <InputBox
        className="Combat__input--larger"
        id="hitDice"
        label="Hit Dice"
        onChange={[Function]}
        type="text"
      />
    </div>
    <div
      className="Combat__tempHitPoints"
    >
      <InputBox
        className="Combat__input--larger"
        id="tempHitPoints"
        label="Temporary Hit Points"
        onChange={[Function]}
        onKeyDown={[Function]}
        type="number"
      />
    </div>
  </div>
</div>
`;
